protobuf/lib/src/protobuf/generated_message.dart

```diff
@@ -506,27 +506,35 @@ abstract class GeneratedMessage {
 
   // Support for generating a read-only default singleton instance.
 
-  static final Map<Function?, Function> _defaultMakers = {};
+  static final Map<Function?, _SingletonMaker<GeneratedMessage>>
+      _defaultMakers = {};
 
   static T Function() _defaultMakerFor<T extends GeneratedMessage>(
-      T Function()? createFn) {
-    return (_defaultMakers[createFn] ??= _createDefaultMakerFor<T>(createFn!))
-        as T Function();
-  }
+          T Function()? createFn) =>
+      _getSingletonMaker(createFn!)._frozenSingletonCreator;
 
-  static T Function() _createDefaultMakerFor<T extends GeneratedMessage>(
-      T Function() createFn) {
-    T? defaultValue;
-    T defaultMaker() {
-      return defaultValue ??= createFn()..freeze();
-    }
+  /// For generated code only.
+  static T $_defaultFor<T extends GeneratedMessage>(T Function() createFn) =>
+      _getSingletonMaker(createFn)._frozenSingleton;
 
-    return defaultMaker;
+  static _SingletonMaker<T> _getSingletonMaker<T extends GeneratedMessage>(
+      T Function() fun) {
+    final oldMaker = _defaultMakers[fun];
+    if (oldMaker != null) {
+      return oldMaker as _SingletonMaker<T>;
```

Comment:
To avoid check in dart2js production code, use an implicit downcast via:
```
    return oldMaker as dynamic;
```

Reply:
Thanks for the suggestion. DOne.